File: src/programmers/level_1/BFS/LV3_가장_먼_노드.py
from collections import deque

def bfs(graph,start,n):
    q=deque()
    q.append((start,0))
    visited=[0 for _ in range(n+1)]
    visited[start]=True
    length=[0 for _ in range(n+1)]
    
    while q:
        current,depth=q.popleft()
        for n in graph[current]:
            if visited[n]:
                continue
            length[n]=depth+1
            visited[n]=True
            q.append((n,depth+1))
    
    length.sort(reverse=True)
    
    max_l=length[0]
    count=0
    for l in length:
        if l==max_l:
            count+=1
        else:
            break
    return count
            

def solution(n, edge):
    
    graph={}
    for (u,v) in edge:
        if u not in graph:
            graph[u]=[]
        if v not in graph:
            graph[v]=[]
        graph[u].append(v)
        graph[v].append(u)
    
    answer = bfs(graph,1,n)
    return answer

File: src/programmers/level_1/BFS/LV3_가장_먼_노드.kt
import kotlin.collections.ArrayDeque

class Solution {
    fun solution(n: Int, edge: Array<IntArray>): Int {
        var answer = 0
        
        val graph=mutableMapOf<Int,MutableList<Int>>()
        
        edge.forEach{
            val a=it[0]
            val b=it[1]
            graph[a]=graph[a]?:mutableListOf<Int>()
            graph[b]=graph[b]?:mutableListOf<Int>()
            graph[a]?.add(b)
            graph[b]?.add(a)
        }
        
        val q=ArrayDeque<Int>()
        val visited=IntArray(n+1){-1}
        visited[1]=0
        q.add(1)
        
        var maxLength=0
        while(q.isNotEmpty()){
            var current=q.removeFirst()
            graph[current]?.forEach{next->
                if(visited[next]<0){
                    visited[next]=visited[current]+1
                    q.add(next)
                }
            }
        }
        visited.sortDescending()
        answer=visited.count{it==visited[0]}
        return answer
    }
}